ghost/ghost/src/db/knex/actor.repository.knex.ts

```diff
@@ -0,0 +1,148 @@
+import {Actor} from '../../core/activitypub/actor.entity';
+import {SettingsCache} from '../../common/types/settings-cache.type';
+
+interface DomainEvents {
+    dispatch(event: unknown): void
+}
+
+export class ActorRepositoryKnex implements ActorRepository {
+    private readonly domainEvents: DomainEvents;
+
+    #defaultActor: Actor;
+
+    constructor(
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        @Inject('knex') private readonly knex: any,
+        @Inject('SettingsCache') settingsCache: SettingsCache,
+        @Inject('DomainEvents') domainEvents: DomainEvents
+    ) {
+        this.#defaultActor = Actor.create({
+            id: ObjectID.createFromHexString('deadbeefdeadbeefdeadbeef'),
+            username: 'index',
+            displayName: settingsCache.get('title'),
+            publicKey: settingsCache.get('ghost_public_key'),
+            privateKey: settingsCache.get('ghost_private_key'),
+            following: [],
+            internal: true
+        });
+        this.domainEvents = domainEvents;
+
+        this.save(this.#defaultActor);
+    }
+
+    #modelToActor(model: any) {
+        return Actor.create({
+            id: ObjectID.createFromHexString(model.id),
+            username: model.data.username,
+            displayName: model.data.displayName,
+            publicKey: model.public_key,
+            privateKey: model.private_key,
+            following: model.data.following,
+            followers: model.data.followers,
+            featured: model.data.features,
+            inbox: model.data.inbox,
+            outbox: model.data.outbox
+        });
+    }
+
+    private async getOneByUsername(username: string) {
+        const row = await this.knex('prototype_activitypub')
+            .leftOuterJoin('prototype_activitypub_actors_keys', 'prototype_activitypub.url', 'prototype_activitypub_actors_keys.actor_id')
+            .where('prototype_activitypub.type', 'Person')
+            .whereJsonObject('prototype_activitypub.data', {username})
+            .first();
+
+        if (!row) {
+            return null;
+        }
+        return this.#modelToActor(row);
+    }
+
+    private async getOneById(id: ObjectID) {
+        const row = await this.knex('prototype_activitypub')
+            .leftOuterJoin('prototype_activitypub_actors_keys', 'prototype_activitypub.url', 'prototype_activitypub_actors_keys.actor_id')
+            .where('prototype_activitypub.type', 'Person')
+            .where('prototype_activitypub.id', id.toHexString())
+            .first();
+
+        if (!row) {
+            return null;
+        }
+        return this.#modelToActor(row);
+    }
+
+    async getOne(identifier: string | ObjectID) {
+        if (identifier instanceof ObjectID) {
+            return await this.getOneById(identifier);
+        } else {
+            return await this.getOneByUsername(identifier);
+        }
+    }
+
+    async save(actor: Actor) {
```

Comment:
@allouis I tried adding a follow via the frontend, and I can see that the workflow for this is via adding it as an activity first, and then only the `acceptFollow` actually add the follow to the array of the entity. But this method is only called from the `postToInbox` method, which I never managed to actually call. I think, I need some more explanation here about how it's expected to work.

Reply:
So the flow is like:

1. Alice creates Follow Activity with Bob as the `object` and sends it to Bob's Inbox
2. Bob receives the Follow Activity, and creates an Accept Activity with the Follow Activity as the `object` and sends it to Alice's Inbox 

So when you're testing follow, you need to have ngrok setup, so that the site you're following is able to send the Accept back to you, which will then get handled in the `postToInbox` method.